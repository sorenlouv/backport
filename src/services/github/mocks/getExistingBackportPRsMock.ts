export const getExistingBackportPRsMock = (repoName: string) => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
