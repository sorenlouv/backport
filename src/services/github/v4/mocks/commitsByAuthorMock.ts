import { DataResponse } from '../fetchCommitsByAuthor';

export const commitsWithPullRequestsMock: DataResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        mergeCommit: {
                          oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 85,
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        baseRefName: 'master',
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
