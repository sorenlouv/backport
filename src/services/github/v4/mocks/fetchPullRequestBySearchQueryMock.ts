import { PullRequestBySearchQueryResponse } from '../fetchPullRequestBySearchQuery';

export const fetchPullRequestBySearchQueryMock: PullRequestBySearchQueryResponse = {
  search: {
    nodes: [
      {
        number: 76492,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'e0f4775b780aada005bdd1774edcceac0ffee006',
          message:
            '[APM] @ts-error -> @ts-expect-error (#76492)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title: '[7.x] [APM] @ts-error -> @ts-expect-error (#76492)',
                  state: 'MERGED',
                  number: 99,
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] @ts-error -> @ts-expect-error (#76492)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 76638,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'fae1e02e0f7a475bf92da05be52a817aa2a84959',
          message:
            '[APM] Avoid negative offset for error marker on timeline (#76638)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:fix',
            },
            {
              name: 'v7.10.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title:
                    '[7.x] [APM] Avoid negative offset for error marker on timeline (#76638)',
                  state: 'MERGED',
                  number: 99,
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Avoid negative offset for error marker on timeline (#76638)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 73120,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'aa68e3b63a4b513294dc58eaf4422e66a0beffb1',
          message:
            '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title:
                    '[7.x] [APM] Add anomaly detection API tests + fixes (#73120)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  number: 99,
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'i18n fixes',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.9] [APM] Add anomaly detection API tests + fixes (#73120)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>\n# Conflicts:\n#\tx-pack/plugins/apm/public/components/app/Settings/anomaly_detection/jobs_list.tsx',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'i18n fix',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72599,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '2fc7112ec27a9f8ded0e2f9e097613721f1179dd',
          message:
            '[APM] Update script with new roles/users (#72599)\n\n* [APM] Update script with new roles/users\r\n\r\n* add log\r\n\r\n* Add validation for http prefix',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'backport:skip',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v8.0.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [],
        },
      },
      {
        number: 72797,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '7e126bfab6a3bfc44f9fa50feecfe22b4634e1a0',
          message: 'Update jobs_list.tsx (#72797)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] Update jobs_list.tsx (#72797)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: 'Update jobs_list.tsx (#72797)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] Update jobs_list.tsx (#72797)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: 'Update jobs_list.tsx (#72797)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 69143,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'd12208c7ea9513529ea1aff42d154db89e3573ff',
          message: '[APM] Fix confusing request/minute viz (#69143)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:enhancement',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Fix confusing request/minute viz (#69143)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Fix confusing request/minute viz (#69143)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 71655,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'f760d8513b0216a73e9a476661f0fb8fb0887a61',
          message: '[APM] Remove watcher integration (#71655)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:deprecation',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Remove watcher integration (#71655)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Remove watcher integration (#71655)\n\n# Conflicts:\n#\tx-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/__test__/esResponse.ts',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72614,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '05ee3da80db34ccf93e7424aa2704c098a1b49fa',
          message: '[APM] Disable flaky rum e2e’s (#72614)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Disable flaky rum e2e’s (#72614)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: '[APM] Disable flaky rum e2e’s (#72614)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Disable flaky rum e2e’s (#72614)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: '[APM] Disable flaky rum e2e’s (#72614)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'Issue',
                },
              },
            },
          ],
        },
      },
      {
        number: 71661,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '51a862988c344b34bd9da57dd57008df12e1b5e5',
          message:
            '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize`',
                  state: 'CLOSED',
                  baseRefName: 'master',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize`',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'Hardcode top traces to 10000 buckets',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message:
                              "Merge branch 'master' into increase-transactionGroupBucketSize",
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.9] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72316,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '511e4543a7828cf0cdb157b88b01352947e0384f',
          message:
            '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Handle ML errors (#72316)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[APM] Fixes error when loading APM without ML read permissions',
                  state: 'CLOSED',
                  baseRefName: 'master',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              'Fixes error when loading APM with an APM ready-only user.',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'Issue',
                },
              },
            },
          ],
        },
      },
    ],
  },
};
