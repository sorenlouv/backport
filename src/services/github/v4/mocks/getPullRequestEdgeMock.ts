import { PullRequestEdge } from '../fetchCommitsByAuthor';

export function getPullRequestEdgeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestEdge {
  return {
    node: {
      labels: {
        nodes: [{ name: 'my-label-a' }],
      },
      mergeCommit: {
        oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      },
      repository: {
        name: 'kibana',
        owner: {
          login: 'elastic',
        },
      },
      number: pullRequestNumber,
      timelineItems: {
        edges: [
          {
            node: {
              source: {
                __typename: 'PullRequest',
                title: timelinePullRequest.title,
                state: 'MERGED' as const,
                commits: {
                  edges: timelinePullRequest.commits.map((message) => ({
                    node: { commit: { message } },
                  })),
                },
                baseRefName: '7.x',
              },
            },
          },
        ],
      },
    },
  };
}
