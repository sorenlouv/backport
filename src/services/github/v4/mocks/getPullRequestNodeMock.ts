import { PullRequestNode } from '../getExistingTargetPullRequests';

export function getPullRequestNodeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestNode {
  return {
    labels: {
      nodes: [{ name: 'my-label-a' }],
    },
    mergeCommit: {
      oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      message: 'my commit message...1',
    },
    repository: {
      name: 'kibana',
      owner: {
        login: 'elastic',
      },
    },
    baseRefName: 'master',
    number: pullRequestNumber,
    timelineItems: {
      edges: [
        {
          node: {
            source: {
              __typename: 'PullRequest',
              title: timelinePullRequest.title,
              state: 'MERGED' as const,
              commits: {
                edges: timelinePullRequest.commits.map((message) => ({
                  node: { commit: { message } },
                })),
              },
              baseRefName: '7.x',
            },
          },
        },
      ],
    },
  };
}
