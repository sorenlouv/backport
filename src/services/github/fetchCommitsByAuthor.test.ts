import { CommitSelected } from './Commit';
import {
  fetchCommitsByAuthor,
  getExistingBackportPRs,
} from './fetchCommitsByAuthor';
import { commitsWithPullRequestsMock } from './mocks/commitsByAuthorMock';
import { getExistingBackportPRsMock } from './mocks/getExistingBackportPRsMock';
import { getDefaultOptions } from '../../test/getDefaultOptions';
import axios from 'axios';
import { getPullRequestEdgeMock } from './mocks/getPullRequestEdgeMock';

const currentUserMock = { user: { id: 'myUserId' } } as const;

describe('fetchCommitsByAuthor', () => {
  describe('when commit has an associated pull request', () => {
    let requestSpy: jasmine.Spy;
    let res: CommitSelected[];
    beforeEach(async () => {
      requestSpy = spyOn(axios, 'post').and.returnValues(
        { data: { data: currentUserMock } },
        { data: { data: commitsWithPullRequestsMock } }
      );

      const options = getDefaultOptions();
      res = await fetchCommitsByAuthor(options);
    });

    it('Should return a list of commits with pullNumber and existing backports', () => {
      expect(res).toEqual([
        {
          sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
          formattedMessage: 'Add ðŸ‘» (2e63475c)',
          existingBackports: [],
        },
        {
          sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
          formattedMessage: 'Add witch (#85)',
          pullNumber: 85,
          existingBackports: [],
        },
        {
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          formattedMessage: 'Add SF mention (#80)',
          pullNumber: 80,
          existingBackports: [{ branch: '6.3', state: 'MERGED' }],
        },
        {
          sha: '3827bbbaf39914eda4f02f6940189844375fd097',
          formattedMessage: 'Add backport config (3827bbba)',
          existingBackports: [],
        },
        {
          sha: '5ea0da550ac191029459289d67f99ad7d310812b',
          formattedMessage: 'Initial commit (5ea0da55)',
          existingBackports: [],
        },
      ]);
    });

    it('should call with correct args to fetch author id', () => {
      expect(requestSpy.calls.argsFor(0)).toMatchSnapshot();
    });

    it('should call with correct args to fetch commits', () => {
      expect(requestSpy.calls.argsFor(1)).toMatchSnapshot();
    });
  });

  describe('existingBackports', () => {
    it('should return existingBackports when repoNames match', async () => {
      const res = await getExistingBackportsByRepoName('kibana', 'kibana');
      expect(res).toEqual([
        {
          existingBackports: [{ branch: '6.3', state: 'MERGED' }],
          formattedMessage: 'Add SF mention (#80)',
          pullNumber: 80,
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
        },
      ]);
    });

    it('should not return existingBackports when repoNames does not match', async () => {
      const res = await getExistingBackportsByRepoName('kibana', 'kibana2');
      expect(res).toEqual([
        {
          existingBackports: [],
          formattedMessage: 'Add SF mention (#80)',
          pullNumber: 80,
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
        },
      ]);
    });
  });

  describe('when a custom github api hostname is supplied', () => {
    it('should be used in gql requests', async () => {
      const requestSpy = spyOn(axios, 'post').and.returnValues(
        { data: { data: currentUserMock } },
        { data: { data: commitsWithPullRequestsMock } }
      );

      const options = getDefaultOptions({
        githubApiBaseUrlV4: 'https://api.github.my-company.com',
      });
      await fetchCommitsByAuthor(options);

      expect(requestSpy.calls.argsFor(0)[0]).toBe(
        'https://api.github.my-company.com'
      );
      expect(requestSpy.calls.argsFor(1)[0]).toBe(
        'https://api.github.my-company.com'
      );
    });
  });
});

describe('getExistingBackportPRs', () => {
  it('should return a result when commit messages match', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message (#1234)'],
      },
    });
    const existingPRs = getExistingBackportPRs(commitMessage, pullRequestEdge);
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });

  it('should not return a result when commit messages do not match', () => {
    const commitMessage = 'my message1 (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message2 (#1234)'],
      },
    });
    const existingPRs = getExistingBackportPRs(commitMessage, pullRequestEdge);
    expect(existingPRs).toEqual([]);
  });

  it('should return a result when commit message matches pull request title and number', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'my message (#1234)',
        commits: ['the actual message'],
      },
    });
    const existingPRs = getExistingBackportPRs(commitMessage, pullRequestEdge);
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });

  it('should not return a result when only pull request title (and not pull number) matches', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'my message (#1235)',
        commits: ['the actual message'],
      },
    });
    const existingPRs = getExistingBackportPRs(commitMessage, pullRequestEdge);
    expect(existingPRs).toEqual([]);
  });

  it('should return a result when first line of a multiline commit message matches', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message (#1234)\n\nsomething else'],
      },
    });
    const existingPRs = getExistingBackportPRs(commitMessage, pullRequestEdge);
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });
});

async function getExistingBackportsByRepoName(
  repoName1: string,
  repoName2: string
) {
  const existingPrsMock = getExistingBackportPRsMock(repoName1);

  spyOn(axios, 'post').and.returnValues(
    { data: { data: currentUserMock } },
    { data: { data: existingPrsMock } }
  );

  const options = getDefaultOptions({
    repoName: repoName2,
  });
  return fetchCommitsByAuthor(options);
}
