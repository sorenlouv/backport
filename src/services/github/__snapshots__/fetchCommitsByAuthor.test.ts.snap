// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch author id 1`] = `
Array [
  "https://api.github.com/graphql",
  Object {
    "query": "
    query getIdByLogin($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
  Object {
    "headers": Object {
      "Authorization": "bearer myAccessToken",
      "Content-Type": "application/json",
    },
  },
]
`;

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch commits 1`] = `
Array [
  "https://api.github.com/graphql",
  Object {
    "query": "
    query getCommitsByAuthorQuery(
      $repoOwner: String!
      $repoName: String!
      $commitsCount: Int!
      $sourceBranch: String!
      $authorId: ID
      $historyPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $commitsCount
                author: { id: $authorId }
                path: $historyPath
              ) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          repository {
                            owner {
                              login
                            }
                            name
                          }
                          number
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      title
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
    "variables": Object {
      "authorId": "myUserId",
      "commitsCount": 10,
      "historyPath": null,
      "repoName": "kibana",
      "repoOwner": "elastic",
      "sourceBranch": undefined,
    },
  },
  Object {
    "headers": Object {
      "Authorization": "bearer myAccessToken",
      "Content-Type": "application/json",
    },
  },
]
`;
